streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -714,13 +696,20 @@ void shutdown(final boolean clean) {
             }
         }
 
-        if (clean && !consumedOffsetsAndMetadataPerTask.isEmpty()) {
-            commitOffsetsOrTransaction(consumedOffsetsAndMetadataPerTask);
+        try {
+            if (clean && !consumedOffsetsAndMetadataPerTask.isEmpty()) {
+                commitOffsetsOrTransaction(consumedOffsetsAndMetadataPerTask);
+            }
+            for (final TaskId taskId : consumedOffsetsAndMetadataPerTask.keySet()) {
+                final Task task = tasks.get(taskId);
+                task.postCommit();
+            }
+        } catch (final RuntimeException e) {
+            firstException.compareAndSet(null, e);
```

Comment:
Why do we do the `try-catch` as outer-layer? In an exception occurs, we should stop looping through the tasks to call `postCommit()` -- is this intended? If yes, why?

Reply:
Well if `commit` throws an exception, then we shouldn't call `postCommit` right? 

Or are you saying if  `commit` succeeds but `postCommit` throws for one task, we should still loop through and try to `postCommit` all the other tasks?

Reply:
Yeah I think if the actual `consumer.commit` call failed, then we should not trigger postCommit for any one.

As for `postCommit`, I think it should never fail (we swallow the IO exception happened, because for non-EOS it is just fine, for EOS we would bootstrap from scratch).

Reply:
I meant the later. And I agree that if `commit` fails, we should not call `postCommit()`.

For failure in `postCommit`: we make assumptions about the current code what seems dangerous (ie, not future prove)? -- IMHO, if `postCommit` fails, we need to close the corresponding task dirty and either recreate it, or rebalance, but we should also continue to call `postCommit()` for all other tasks?

Reply:
I see. Then I think it makes sense to always attempt to write the checkpoint/call `postCommit` for a task that was successfully committed, regardless of whether something went wrong during `postCommit` with a different task 

And I agree, we should not make assumptions about the current code not throwing, unless it's explicitly in the contract of the method that it will never throw (which is not the case for `postCommit`

Reply:
Sounds good, in that case the nested try-catch would be necessary.